# **Phase 3: Data Cleaning**

In [1]:
# Loading file in Python for data cleaning.
import pandas as pd
df = pd.read_csv('Pakistan_Ecommerce_Dataset_for_Cleaning.csv')
df

/tmp/ipykernel_12479/226465248.py:3: DtypeWarning: Columns (1,2,3,7,8,9,11,12,13,14,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Pakistan_Ecommerce_Dataset_for_Cleaning.csv')


,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,discount_amount,payment_method,Working Date,BI Status,MV,Year,Month,Customer Since,M-Y,FY,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,0.0,cod,7/1/2016,#REF!,"1,950",2016.0,7.0,2016-7,7-2016,FY17,1.0,NaN,NaN,NaN,NaN,NaN
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,\N,0.0,cod,7/1/2016,Gross,240,2016.0,7.0,2016-7,7-2016,FY17,2.0,NaN,NaN,NaN,NaN,NaN
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,\N,0.0,cod,7/1/2016,Gross,"2,450",2016.0,7.0,2016-7,7-2016,FY17,3.0,NaN,NaN,NaN,NaN,NaN
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,R-FSD-52352,300.0,cod,7/1/2016,Net,360,2016.0,7.0,2016-7,7-2016,FY17,4.0,NaN,NaN,NaN,NaN,NaN
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,\N,0.0,cod,7/1/2016,Valid,"1,110",2016.0,7.0,2016-7,7-2016,FY17,5.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1048570,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1048571,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1048572,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1048573,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [2]:
# Removing null values in the item_id column of the dataset.
df = df.dropna(subset=['item_id']).copy()
df

,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,discount_amount,payment_method,Working Date,BI Status,MV,Year,Month,Customer Since,M-Y,FY,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,0.0,cod,7/1/2016,#REF!,"1,950",2016.0,7.0,2016-7,7-2016,FY17,1.0,NaN,NaN,NaN,NaN,NaN
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,\N,0.0,cod,7/1/2016,Gross,240,2016.0,7.0,2016-7,7-2016,FY17,2.0,NaN,NaN,NaN,NaN,NaN
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,\N,0.0,cod,7/1/2016,Gross,"2,450",2016.0,7.0,2016-7,7-2016,FY17,3.0,NaN,NaN,NaN,NaN,NaN
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,R-FSD-52352,300.0,cod,7/1/2016,Net,360,2016.0,7.0,2016-7,7-2016,FY17,4.0,NaN,NaN,NaN,NaN,NaN
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,\N,0.0,cod,7/1/2016,Valid,"1,110",2016.0,7.0,2016-7,7-2016,FY17,5.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
584519,905204.0,cod,8/28/2018,WOFSCE5AE00357AECDE,699.0,1.0,849.0,100562385,Women's Fashion,NaN,0.0,cod,8/28/2018,Valid,699,2018.0,8.0,2018-8,8-2018,FY19,115320.0,NaN,NaN,NaN,NaN,NaN
584520,905205.0,processing,8/28/2018,MATHUA5AF70A7D1E50A,35599.0,1.0,35899.0,100562386,Mobiles & Tablets,NaN,0.0,bankalfalah,8/28/2018,Gross,"35,599",2018.0,8.0,2018-8,8-2018,FY19,115326.0,NaN,NaN,NaN,NaN,NaN
584521,905206.0,processing,8/28/2018,MATSAM5B6D7208C6D30,129999.0,2.0,652178.0,100562387,Mobiles & Tablets,NaN,0.0,bankalfalah,8/28/2018,Gross,"259,998",2018.0,8.0,2018-7,8-2018,FY19,113474.0,NaN,NaN,NaN,NaN,NaN
584522,905207.0,processing,8/28/2018,MATSAM5B1509B4696EA,87300.0,2.0,652178.0,100562387,Mobiles & Tablets,NaN,0.0,bankalfalah,8/28/2018,Gross,"174,600",2018.0,8.0,2018-7,8-2018,FY19,113474.0,NaN,NaN,NaN,NaN,NaN


In [3]:
# Deleting all unnamed columns from the dataset.
df = df.loc[:, ~df.columns.str.contains('^Unnamed')].copy()

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 584524 entries, 0 to 584523
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   item_id                584524 non-null  float64
 1   status                 584509 non-null  object 
 2   created_at             584524 non-null  object 
 3   sku                    584504 non-null  object 
 4   price                  584524 non-null  float64
 5   qty_ordered            584524 non-null  float64
 6   grand_total            584524 non-null  float64
 7   increment_id           584524 non-null  object 
 8   category_name_1        584360 non-null  object 
 9   sales_commission_code  447346 non-null  object 
 10  discount_amount        584524 non-null  float64
 11  payment_method         584524 non-null  object 
 12  Working Date           584524 non-null  object 
 13  BI Status              584524 non-null  object 
 14   MV                    584524 non-null  o

## **Data Cleaning & Structural Integrity**

In this phase, the raw dataset was streamlined to remove technical noise and non-functional records identified during the initial loading phase.

### **3.1 Global Data Reduction (Removing Ghost Records)**
Upon initial inspection, the dataset contained **1,048,575 rows**, but approximately **44%** of these were completely empty (Null).
* **Action:** I performed a targeted drop of all rows where **item_id** was missing. 
* **Logic:** Since **item_id** is the primary identifier for a transaction, any record without it is considered a "Ghost Row" and provides no analytical value.
* **Technical Note:** After dropping the rows, I created a deep copy of the dataframe using **.copy()** to avoid memory fragmentation and **SettingWithCopyWarnings** in future steps.
* **Outcome:** The dataset was reduced to **584,524 rows**, which represents the actual transactional data.

### **3.2 Removing Irrelevant Metadata (Unnamed Columns)**
The original file contained 5 "Unnamed" columns (labeled 21 to 24/25) that carried no headers and zero data.
* **Action:** These columns were dropped to optimize memory usage and focus only on relevant business variables.
* **Current Shape:** The dataset now stands at **584,524 rows and 21 columns**.

### **3.3 Roadmap: Individual Column Audit (Post-Mortem)**
To ensure the highest data quality, I am now transitioning from "Global Cleaning" to "Granular Auditing." Each column will be inspected individually for:
1. **Data Type Correction:** Ensuring IDs are integers, prices are floats, and dates are in proper Datetime format.
2. **Value Standardization:** Fixing inconsistent labels in categorical columns like **status**.
3. **Logic Validation:** Ensuring numerical values (price, quantity) are logical and positive.

In [5]:
# Converting data types of ID columns (item_id & customerID) to integer data type.
# Converting 'item_id' to integer data type.
df['item_id'] = df['item_id'].astype('int64')

# Checking if there any missing values in 'Customer ID' column and then converting it to integer data type.
df['Customer ID'] = df['Customer ID'].fillna(0).astype('int64')

# Verification
print(df[['item_id', 'Customer ID']].dtypes)
print(df[['item_id', 'Customer ID']])

item_id        int64
Customer ID    int64
dtype: object
        item_id  Customer ID
0        211131            1
1        211133            2
2        211134            3
3        211135            4
4        211136            5
...         ...          ...
584519   905204       115320
584520   905205       115326
584521   905206       113474
584522   905207       113474
584523   905208       113474

[584524 rows x 2 columns]


In [6]:
# 1.Cleaning and standardizing 'Status' column.
df['status'] = df['status'].replace(['\\N', 'nan'], 'unknown')
df['status'] = df['status'].fillna('unknown').str.lower()

# 2.Cleaning and standardizing 'SKU' column.
df['sku'] = df['sku'].fillna('unknown').str.lower()

# 3. Cleaning and standardizing 'Payment Method' column.
df['payment_method'] = df['payment_method'].fillna('unknown').str.lower()

# 4. Checking 'null' in 'category_name_1' column.
print(f"Nulls in Category: {df['category_name_1'].isnull().sum()}")
# Removing 'null', if any exists.
df['category_name_1'] = df['category_name_1'].fillna('unknown').str.lower()

# Verification of 'SKU' and 'Payment Method' columns.
print(f"Nulls in SKU: {df['sku'].isnull().sum()}")
print(f"Nulls in Payment Method: {df['payment_method'].isnull().sum()}")
print("\nFirst 5 SKUs:", df['sku'].head())

Nulls in Category: 164
Nulls in SKU: 0
Nulls in Payment Method: 0

First 5 SKUs: 0                                    kreations_yi 06-l
1    kcc_buy 2 frey air freshener & get 1 kasual bo...
2                                   ego_up0017-999-mr0
3                                       kcc_krone deal
4                                          bk7010400ag
Name: sku, dtype: object


### **3.4 Granular Data Transformation (Post-Mortem Audit)**

After restructuring the dataframe, I performed an individual column audit to align data types with business logic and handle hidden inconsistencies.

#### **A. Identifier Standardization**
* **Columns:** **item_id**, **Customer ID**
* **Action:** Converted from **float64** to **int64**. 
* **Reasoning:** Transactional and Customer IDs are discrete whole numbers. Removing decimals ensures indexing integrity and reduces the memory footprint.

#### **B. Categorical Consistency & Missing Value Imputation**
* **Columns:** **status**, **sku**, **payment_method**, **category_name_1**
* **Action:** 
    1. Identified technical artifacts **(\\N, nan)** and actual nulls.
    2. Imputed all missing entries with a generic **'unknown'** label to maintain dataset completeness.
    3. Converted all strings to **lowercase** for uniform grouping.
* **Reasoning:** SKU and Status are critical for product-level analysis. Standardization prevents data fragmentation and ensures 100% record availability for categorical aggregation.

In [7]:
# 1. Converting 'created_at' and 'Working Date' to proper datetime
df['created_at'] = pd.to_datetime(df['created_at'])
df['Working Date'] = pd.to_datetime(df['Working Date'])
# Creating 'day_name' & 'month_name' columns for EDA.
df['day_name'] = df['created_at'].dt.day_name()
df['month_name'] = df['created_at'].dt.month_name()

# 2. Converting 'Year' and 'Month' to integer.
df['Year'] = df['Year'].astype('int64')
df['Month'] = df['Month'].astype('int64')

# Verification
print(df[['created_at', 'Working Date', 'Year', 'Month']].dtypes)
print(df[['created_at', 'day_name', 'month_name']].head())
print("\nSample Dates:\n", df[['created_at', 'Working Date', 'Year', 'Month']].head())

created_at      datetime64[ns]
Working Date    datetime64[ns]
Year                     int64
Month                    int64
dtype: object
  created_at day_name month_name
0 2016-07-01   Friday       July
1 2016-07-01   Friday       July
2 2016-07-01   Friday       July
3 2016-07-01   Friday       July
4 2016-07-01   Friday       July

Sample Dates:
   created_at Working Date  Year  Month
0 2016-07-01   2016-07-01  2016      7
1 2016-07-01   2016-07-01  2016      7
2 2016-07-01   2016-07-01  2016      7
3 2016-07-01   2016-07-01  2016      7
4 2016-07-01   2016-07-01  2016      7


#### **C. Feature Engineering for Time-Series Analysis**
* **Columns Created:** **day_name**, **month_name**
* **Action:** Extracted categorical time components from the **created_at** datetime object.
* **Reasoning:** To address key business questions regarding peak sales periods. These features allow us to visualize order frequency by day of the week and identify seasonal trends across different months.

In [8]:
# 1. Total rows before dropping.
print(f"Rows before dropping: {len(df)}")

# 2. Dropping rows where 'Grand Total' or 'Discount Amount' is negative.
df = df[df['grand_total'] >= 0].copy()
df = df[df['discount_amount'] >= 0].copy()

# 3. Converting 'Quantity Orders' to integer.
df['qty_ordered'] = df['qty_ordered'].astype('int64')

# 4. Final verification
print(f"Rows after dropping negative values: {len(df)}")
print(f"Total rows removed: 584524 - {len(df)}")

# 5. Check dtypes one last time
print("\nFinal Financial Dtypes:")
print(df[['price', 'qty_ordered', 'grand_total', 'discount_amount']].dtypes)

Rows before dropping: 584524


Rows after dropping negative values: 584445
Total rows removed: 584524 - 584445

Final Financial Dtypes:
price              float64
qty_ordered          int64
grand_total        float64
discount_amount    float64
dtype: object


#### **D. Quantitative Integrity & Outlier Treatment**
* **Columns:** **price**, **qty_ordered**, **grand_total**, **discount_amount**
* **Action:** 
    1. Identified and **dropped 76 rows** with negative **grand_total**.
    2. Identified and **dropped 3 rows** with negative **discount_amount**.
    3. Converted **qty_ordered** to **int64** for logical consistency.
* **Reasoning:** Financial metrics cannot be negative in a standard sales context. Dropping these records ensures that aggregate calculations (Sum, Mean) reflect actual business revenue without being skewed by data entry errors or logical outliers.

### **3.5 Final Data Cleaning & Structural Integrity Log**

The data cleaning phase is now concluded. Every column has been audited for technical artifacts, logical consistency, and optimal data typing.

| Category | Column(s) | Transformation Action | Final Dtype |
| :--- | :--- | :--- | :--- |
| **Identifiers** | **item_id**, **Customer ID** | Handled NaNs; Cast to discrete whole numbers. | **int64** |
| **Categorical** | **status**, **sku**, **payment_method** | Standardized nulls to 'unknown'; Lowercased all strings. | **object** |
| **Temporal** | **created_at**, **Working Date** | Converted raw strings to functional datetime objects. | **datetime64** |
| **Temporal** | **Year**, **Month** | Optimized from float to integer for grouping. | **int64** |
| **Financial** | **price**, **grand_total** | Audited for business logic; Removed negative outliers. | **float64** |
| **Quantity** | **qty_ordered** | Validated positive bounds; Cast to integer. | **int64** |

**Final Dataset Metrics:**
* **Original Records:** 1,048,575
* **Cleaned Records:** 584,445 (approx.)
* **Data Health:** 100% Non-Null in core analytical columns.

In [9]:
# Saving file to use in other notebooks.
df.to_csv('Pakistan_Ecommerce_Dataset_for_EDA.csv', index=False)
print('File successfully saved for EDA.')

File successfully saved for EDA.
